src/rust/engine/src/fs.rs

```diff
@@ -636,26 +638,42 @@ impl fmt::Debug for Snapshot {
 }
 
 /**
- * A facade for the snapshot directory, which is currently thrown away at the end of each run.
+ * A facade for the snapshot directory, which lives under the pants workdir.
  */
 pub struct Snapshots {
-  temp_dir: TempDir,
-  next_temp_id: atomic::AtomicUsize,
+  snapshots_dir: PathBuf,
+  snapshots_tmpdir: TempDir,
+  next_temp_id: atomic::AtomicUsize
 }
 
 impl Snapshots {
-  pub fn new() -> Result<Snapshots, io::Error> {
+  pub fn new(snapshots_dir: PathBuf) -> Result<Snapshots, String> {
+    let mut snapshots_tmpdir_base = snapshots_dir.clone();
+    snapshots_tmpdir_base.push(".tmp");
+
+    let snapshots_tmpdir = util::safe_create_tmpdir_in(&snapshots_tmpdir_base, "snapshots")?;
+
     Ok(
       Snapshots {
-        // TODO: see https://github.com/pantsbuild/pants/issues/4299
-        temp_dir: TempDir::new("snapshots")?,
-        next_temp_id: atomic::AtomicUsize::new(0),
+        snapshots_dir: snapshots_dir,
+        snapshots_tmpdir: snapshots_tmpdir,
+        next_temp_id: atomic::AtomicUsize::new(0)
       }
     )
   }
 
-  pub fn path(&self) -> &Path {
-    self.temp_dir.path()
+  pub fn snapshot_path(&self) -> &Path {
+    self.snapshots_dir.as_path()
+  }
+
+  fn tmp_path(&self) -> &Path {
+    // N.B. Sometimes, in e.g. a `./pants clean-all test ...` the snapshot tempdir created at
+    // the beginning of a run can be removed out from under us by e.g. the clean-all task. Here,
+    // we double check existence of the dir (by creating it, to avoid an extra stat) when the path
+    // is accessed.
+    let path = self.snapshots_tmpdir.path();
+    util::safe_create_dir_all(&path).expect("Could not create snapshot temp dir.");
```

Comment:
I think I'd prefer to see this at the location that's writing the temp file, rather than as a sideeffect of getting the directory.